File: C5_8.c
#include <stdio.h>

void rule(int, int, int);

int main(void)
{
    int i = 8, j = 3;
    rule(0, 8, 3);
    return 0;
}

void rule(int l, int r, int h)
{
    int m = (l + r) / 2;
    if (h > 0)
    {
        printf("%d:%d ", m, h);
        rule(l, m, h - 1);
        //printf("%d:%d ", m, h);
        rule(m, r, h - 1);
    }
}

File: C5.8.py
def rule(l, r, h):
    m = l + (r-l)/2
    if h > 0:
        result[m] = h
        rule(l, m, h-1)

        rule(m, r, h-1)




result = {}
if __name__ == "__main__":
    l, r, h = 0, 128, 4
    rule(l, r, h)
    for i in (sorted(result.keys())):
        print("%d: %d " %(i, result[i]))


